vignettes/programming_strategy.Rmd

```diff
@@ -89,6 +89,8 @@ is restricted, otherwise this step is skipped.
 on `BASE` or based on `ANRIND`. It should not be implemented within one function as the algorithms are completely different.
 If `BASE` is used, the values are categorized while if `ANRIND` is used, the values are merged from the baseline observation.
 
+  * If developers find the need to use or create *environment* objects to achieve flexibility, use the `admiral_environment` environment object created in `admiral_environment.R`. All objects which are stored in this environment must be documented in `admiral_environment.R`. An equivalent environment object and `.R` file exist for admiraldev as well. For more details how environments work, please refer to this [chapter](https://r-pkgs.org/data.html#sec-data-state) in the latest edition of the R Packages book.
```

Comment:
Can you elaborate a little more on what you mean by:
"To avoid that other packages store datasets in the `admiraldev_environment`? "

Also still trying to wrap my head around a couple more things, and as I read the advanced R chapter on this topic its helping. _Currently_ the environment `.datasets` is being "created" in both packages (`admiral/R/duplicates.R` and `admiraldev/R/datasets.R`) through the line:

`
.datasets <- new.env(parent = emptyenv())
`

But "exists" as a singular environment object to point to:

- `duplicates` created from the admiral function `signal_duplicate_records()`
- `one_to_many` and `many_to_one` created from the admiraldev function `assert_one_to_one()` inside admiral's `derive_param_tte()`

Is the intuition that `duplicates` would exist `admiral_environment` and `one_to_many` and `many_to_one` would exist `admiraldev_environment` ?

Reply:
By "To avoid that other packages store datasets in the `admiraldev_environment`" I mean that `set_dataset()` looks like a function which could be used by any function outside admiraldev. Assume for example I would implement a TA specific assertion in admiralonco where I want to provide a dataset of problematic observations. I may call `set_dataset(invalidobs, name = "dataissues")` in this assertion. This would work until another developer in a different package has a similar idea and calls `set_dataset(invalidobs, name = "dataissues")` in a function as well. Then both functions would store their data in `admiraldev_environment$dataissues` and one function would overwrite the data from the other function.

Objects created outside of functions like `.dataset` are created in the package environment. I.e., the `.dataset` environment from `admiral/R/duplicates.R` is created in the `admiral` environment while the `.dataset` environment from `admiraldev/R/datasets.R` is created in the `admiraldev` environment.
I think `duplicates` should be in `admiral_environment` because it is created by the admiral function `signal_duplicate_records()` and `one_to_many` and `many_to_one` should be in `admiraldev_environment` because they are created by the admiraldev function `assert_one_to_one()`.
In principle we could name all environments `admiral_environment`, e.g., use
```
admiral_environment <- new.env(parent = emptyenv())
```
in admiraldev. But I think `admiraldev_environment` is clearer.